hathor/manager.py

```diff
@@ -633,59 +719,99 @@ def propagate_tx(self, tx: BaseTransaction, fails_silently: bool = True) -> bool
         else:
             tx.storage = self.tx_storage
-        return self.on_new_tx(tx, fails_silently=fails_silently)
+
+        return self.on_new_tx(tx, fails_silently=fails_silently, propagate_to_peers=True)
 
     @cpu.profiler('on_new_tx')
     def on_new_tx(self, tx: BaseTransaction, *, conn: Optional[HathorProtocol] = None,
                   quiet: bool = False, fails_silently: bool = True, propagate_to_peers: bool = True,
-                  skip_block_weight_verification: bool = False) -> bool:
-        """This method is called when any transaction arrive.
-
-        If `fails_silently` is False, it may raise either InvalidNewTransaction or TxValidationError.
-
-        :return: True if the transaction was accepted
-        :rtype: bool
+                  skip_block_weight_verification: bool = False, partial: bool = False) -> bool:
+        """ New method for adding transactions or blocks that steps the validation state machine.
         """
         assert tx.hash is not None
-        if self.state != self.NodeState.INITIALIZING:
-            if self.tx_storage.transaction_exists(tx.hash):
-                if not fails_silently:
-                    raise InvalidNewTransaction('Transaction already exists {}'.format(tx.hash_hex))
-                self.log.debug('on_new_tx(): Transaction already exists', tx=tx.hash_hex)
-                return False
+        if self.tx_storage.transaction_exists(tx.hash):
+            if not fails_silently:
+                raise InvalidNewTransaction('Transaction already exists {}'.format(tx.hash_hex))
+            self.log.warn('on_new_tx(): Transaction already exists', tx=tx.hash_hex)
+            return False
 
-        if self.state != self.NodeState.INITIALIZING or self._full_verification:
+        if tx.timestamp - self.reactor.seconds() > settings.MAX_FUTURE_TIMESTAMP_ALLOWED:
+            if not fails_silently:
+                raise InvalidNewTransaction('Ignoring transaction in the future {} (timestamp={})'.format(
+                    tx.hash_hex, tx.timestamp))
+            self.log.warn('on_new_tx(): Ignoring transaction in the future', tx=tx.hash_hex,
+                          future_timestamp=tx.timestamp)
+            return False
+
+        tx.storage = self.tx_storage
+
+        try:
+            metadata = tx.get_metadata()
+        except TransactionDoesNotExist:
+            if not fails_silently:
+                raise InvalidNewTransaction('missing parent')
+            self.log.warn('on_new_tx(): missing parent', tx=tx.hash_hex)
+            return False
+
+        if metadata.validation.is_invalid():
+            if not fails_silently:
+                raise InvalidNewTransaction('previously marked as invalid')
+            self.log.warn('on_new_tx(): previously marked as invalid', tx=tx.hash_hex)
+            return False
+
+        if not partial or (metadata.validation.is_fully_connected() or tx.can_validate_full()):
```

Comment:
I think this `if` here deserves a comment with a short explanation. It was not easy at first to understand the usage of `partial` parameter

Reply:
I added a short explanation.